OpenSearch Extensibility Roadmap (#1455)
opensearch-project/project-website: _posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown

---
layout: post
title:  "Technical roadmap: OpenSearch extensibility"
authors:
  - dblock
date:   2023-04-03
categories:
  - technical-post
meta_keywords: OpenSearch extensibility, extensions framework, OpenSearch plugins, OpenSearch Dashboards
meta_description: Learn how the OpenSearch Project is replacing its plugin framework with a new catalog of extensions to help users and developers overcome extensibility challenges.
excerpt: The OpenSearch plugin framework presents a number of challenges for users and developers. To solve these, we've embarked on a journey to replace the plugin mechanism with extensions.
---

The primary reason users choose OpenSearch is the wide range of use cases they can address with its features, such as search or log analytics. Thus, we aim to make the OpenSearch Project the preferred platform for builders by creating a vibrant and deeply integrated ecosystem of projects, features, content packs, integrations, and tools that can be found quickly, installed securely, combined to solve problems, and monetized by many participants. 

> The existing mechanism used to extend OpenSearch and OpenSearch Dashboards is a plugin framework. It provides a useful way to extend functionality, particularly when the new functionality needs access to a significant number of internal APIs. However, the plugin framework presents a number of challenges for users and developers in the areas of administration, dependency management, security, availability, scalability, and developer velocity. To begin solving these, we've embarked on a journey to replace the OpenSearch plugin mechanism with a new catalog of _extensions_. We plan to ship two new SDKs for OpenSearch and OpenSearch Dashboards, and then launch a catalog of extensions.

Reviewer: ```suggestion
The existing mechanism used to extend OpenSearch and OpenSearch Dashboards is a plugin framework. It provides a useful way to extend functionality, particularly when the new functionality needs access to a significant number of internal APIs. However, the plugin framework presents a number of challenges for users and developers in the areas of administration, dependency management, security, availability, scalability, and developer velocity. To begin solving these, we've embarked on a journey to replace the OpenSearch plugin mechanism with a new catalog of _extensions_. We plan to ship two new SDKs for OpenSearch and OpenSearch Dashboards and then launch a catalog of extensions.
```

In this blog post, we'll introduce the concept of extensions and outline some proposed projects in this area.

## Introducing extensions

> From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model of building new features in OpenSearch. Technically, extensions are a simple evolution of plugins --- or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashboards SDK, and works with multiple versions of OpenSearch/Dashboards. We aim for the existence of many more extensions than there are plugins today, written by many more developers. Then we will sunset plugins.

Reviewer: ```suggestion
From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model for building new features in OpenSearch. Technically, extensions are a simple evolution of plugins---or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashboards SDK and works with multiple versions of OpenSearch/Dashboards. We aim for the existence of many more extensions than there are plugins today, written by many more developers. Then we will sunset plugins.
```

> We want for extensions to become the preferred mechanism for providing functionality in OpenSearch/Dashboards. Having multiple competing implementations will produce extensions with high performance, improved security, and versatile features for all categories of users, administrators and developers.

Reviewer: ```suggestion
We want for extensions to become the preferred mechanism for providing functionality in OpenSearch/Dashboards. Having multiple competing implementations will produce extensions with high performance, improved security, and versatile features for all categories of users, administrators, and developers.
```

## Extensions roadmap

The following proposed projects are in chronological order, but many of them can be accomplished in parallel.

### Provide experimental SDKs and move extensibility concerns out of cores

> First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK, and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards without having to upgrade an installed extension.

Reviewer: ```suggestion
First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards without having to upgrade an installed extension.
```

> The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboards >= 2.3.0` (2.3, 2.4, ... 3.0, and so on) or `~> 2.5` (any 2.x after 2.5). The SDK will provide support for integration testing against broad ranges of OpenSearch/Dashboards versions. It can begin selecting common functionality that all plugins may need, such as storing credentials or saving objects, and be strongly opinionated about what constitutes a semver-compatible extension point for OpenSearch/Dashboards. Instead of importing a transitive dependency (e.g. `oui`), developers import an SDK namespace (e.g. `sdk/ui`).

Reviewer: ```suggestion
The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboards >= 2.3.0` (2.3, 2.4, ... 3.0, and so on) or `~> 2.5` (any 2.x after 2.5). The SDK will provide support for integration testing against broad ranges of OpenSearch/Dashboards versions. It can begin selecting common functionality that all plugins may need, such as storing credentials or saving objects, and be strongly opinionated about what constitutes a semver-compatible extension point for OpenSearch/Dashboards. Instead of importing a transitive dependency (for example, `oui`), developers import an SDK namespace (for example, `sdk/ui`).
```

### Add security support for extensions in the OpenSearch/Dashboards cores

In the plugin model, security is also a plugin. This means security can be optional, which makes it difficult for plugins to build features such as field- and document-level security (FLS/DLS), which perform data access securely. Each plugin must implement data access checks independently and correctly, which has proven to be difficult (see [security #1895](https://github.com/opensearch-project/security/issues/1895) for some examples). A simpler and more secure implementation would make all operations inside OpenSearch permissible, no matter the source, syncing access checks to the lower levels. The SDKs move extensibility concerns out of the monoliths. They will have brand-new APIs, presenting the most opportune time to require a security context in all OpenSearch API calls. It should, of course, still be possible to disable security as needed.

> We plan to add authentication mechanisms to the OpenSearch core (every API call or new thread/process will carry an identity), and perform authorization when accessing data at the level of these APIs in a way that is backward compatible with the Security plugin. Authorization check will be enabled using the Security plugin for core APIs exposed in the SDK, and there will be no changes required in plugins to ensure backward compatibility.

Reviewer: ```suggestion
We plan to add authentication mechanisms to the OpenSearch core (every API call or new thread/process will carry an identity) and perform authorization when accessing data at the level of these APIs in a way that is backward compatible with the Security plugin. Authorization checks will be enabled using the Security plugin for core APIs exposed in the SDK, and there will be no changes required in plugins to ensure backward compatibility.
```

### Replace plugins in the OpenSearch/Dashboards distribution

One of the main reasons for the existence of the default OpenSearch distribution is that it provides a set of secure, signed binaries along with a rich set of features (plugins). We have invested in a large automation effort in the open-source distribution to make this process safe and repeatable. However, producing this distribution still requires significant logistical and technical coordination, plus the semi-automated labor of incrementing versions, adding components to manifests, and tracking unstable upstreams. The toughest challenge to overcome is developers having to develop, build, and test plugins against moving targets of non-stable versions of the OpenSearch/Dashboards cores.

We aim for extensions to ship independently, either more or less often than the official distribution. Users should be able to upgrade OpenSearch clusters much more easily because they won’t need to upgrade installed extensions. Additionally, with fewer versions of released extensions containing no new features, there will be less security patching.

> For each existing plugin that ships with the OpenSearch distribution, we will need to design a technical path forward, but this phase will present no change in user experience for the default OpenSearch/Dashboards distribution. First, we will design and implement interfaces that need to be exported via the OpenSearch/Dashboards SDK. This presents an opportunity to redesign extension points to be simpler and more coherent and an opportunity to refactor classes in OpenSearch/Dashboards. Plugins remove the dependency on OpenSearch/Dashboards and add a dependency on the SDK, reimplement the calls, and are now available as extensions. Second, to migrate the entire Security plugin into the core, we will need to add support for authorization in REST handlers, implement authorized request forwarding in the REST layer, add support for asynchronous operations and background tasks, and add system index support to allow extensions to reserve system indexes. Finally, the distribution mechanisms can begin picking up the latest available version of an extension and releasing together instead of rebuilding everything.

Reviewer: ```suggestion
For each existing plugin that ships with the OpenSearch distribution, we will need to design a technical path forward, but this phase will present no change in user experience for the default OpenSearch/Dashboards distribution. First, we will design and implement interfaces that need to be exported via the OpenSearch/Dashboards SDK. This presents an opportunity to redesign extension points to be simpler and more coherent and an opportunity to refactor classes in OpenSearch/Dashboards. Plugins remove the dependency on OpenSearch/Dashboards and add a dependency on the SDK, reimplement the calls, and are then available as extensions. Second, to migrate the entire Security plugin into the core, we will need to add support for authorization in REST handlers, implement authorized request forwarding in the REST layer, add support for asynchronous operations and background tasks, and add system index support to allow extensions to reserve system indexes. Finally, the distribution mechanisms can begin picking up the latest available version of an extension and releasing together instead of rebuilding everything.
```

> For each existing plugin that ships with the OpenSearch distribution, we will need to design a technical path forward, but this phase will present no change in user experience for the default OpenSearch/Dashboards distribution. First, we will design and implement interfaces that need to be exported via the OpenSearch/Dashboards SDK. This presents an opportunity to redesign extension points to be simpler and more coherent and an opportunity to refactor classes in OpenSearch/Dashboards. Plugins remove the dependency on OpenSearch/Dashboards and add a dependency on the SDK, reimplement the calls, and are now available as extensions. Second, to migrate the entire Security plugin into the core, we will need to add support for authorization in REST handlers, implement authorized request forwarding in the REST layer, add support for asynchronous operations and background tasks, and add system index support to allow extensions to reserve system indexes. Finally, the distribution mechanisms can begin picking up the latest available version of an extension and releasing together instead of rebuilding everything.

Reviewer: I would include a noun or pronoun between "releasing" and "together" in the last sentence. Releasing what together?

### OpenSearch Dashboards cohesion through interfaces

> OpenSearch Dashboards is a single product that includes the core platform (e.g. the application chrome, data, and saved objects APIs), native plugins (e.g. home, discover, dev tools, and stack management), and feature plugins (e.g. observability, anomaly detection, maps, and alerting).

Reviewer: ```suggestion
OpenSearch Dashboards is a single product that includes the core platform (for example, the application chrome, data, and saved objects APIs), native plugins (for example, Home, Discover, Dev Tools, and Stack Management), and feature plugins (for example, Observability, Anomaly Detection, Maps, and Alerting).
```

### Extensible document parsing

JSON is, by far, the most popular input format for OpenSearch. JSON is humanly readable, so it is pretty easy to test and use for development. Additionally, the OpenSearch ecosystem is built around JSON with most benchmarks written in JSON and ingest connectors supporting JSON. However, JSON is much slower and more space-consuming than most binary formats, thus swapping JSON for another type may yield significant performance gains. We would like to make OpenSearch input formats extensible so that it is easier to add and test more formats. This was proposed in [OpenSearch #4559](https://github.com/opensearch-project/OpenSearch/issues/4559).

### Security isolation

> With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (e.g. Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime, and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypted at rest.

Reviewer: ```suggestion
With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (for example, Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypted at rest.
```

> With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (e.g. Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime, and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypted at rest.

Reviewer: Last sentence: Either "We can further extend security across new boundaries" or "We can further secure [noun] across new boundaries".

### Aggregations as an extension

As we work toward separating search aggregations and compute, we will refactor and extract interface and SDK components for compute separately from the search phase lifecycle. Simple search aggregations, which default to the current single-pass data aggregation implementation, may still be supported for basic search use cases. Separating the compute framework through extensions and a developer SDK enables drop-in replacements from more evolved data analytics systems, such as Spark. By separating the search aggregation framework through extensions and a developer SDK, third-party contributors can leverage new kinds of search aggregation support, such as principal component analysis (PCA) or stratified sampling pipeline aggregations.

### Cluster manager as an extension

We would like to refactor and extract cluster management interfaces to remove the existing cluster manager node limit of ~200. For large deployments, we believe we can provide alternate cluster manager implementations that have a different scalability and availability model.

### Offloading background tasks to extensions

> A number of background operations currently run asynchronously in an OpenSearch cluster, including Lucene segment merges, data tier migrations (e.g. hot to warm to cold), and most operations performed by Index State Management (ISM). These can be offloaded to dedicated or isolated compute instances through the extensions mechanism, improving cluster scalability and availability.

Reviewer: ```suggestion
A number of background operations currently run asynchronously in an OpenSearch cluster, including Lucene segment merges, data tier migrations (for example, hot to warm to cold), and most operations performed by Index State Management (ISM). These can be offloaded to dedicated or isolated compute instances through the extensions mechanism, improving cluster scalability and availability.
```

### Communication protocols as extensions

By making communication protocols extensible, we can experiment with more performant implementations such as [GRPC](https://grpc.io/) or no- or low-garbage-collection implementations (e.g. [Netty](https://netty.io/)) without having to modify the core engine. Check out [opensearch-sdk-java #414](https://github.com/opensearch-project/opensearch-sdk-java/issues/414#issuecomment-1471010441), which prototypes protobuf serialization.

## Help wanted

> This blog post reflects some of our current project plans. Like all plans, they may change as we make progress, and we would love your help! The best way to start is by checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java). Try to implement a trivial extension on top of it, or help us port an existing plugin. You could also pickup one of the issues labelled "[good first issue](https://github.com/opensearch-project/opensearch-sdk-java/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)" in that project, or start with one of the future ideas we mentioned above. As always, please let us know how we can help by opening new issues or posting to [the forums](https://forum.opensearch.org/).

Reviewer: ```suggestion
This blog post reflects some of our current project plans. Like all plans, they may change as we make progress, and we would love your help! The best way to start is by checking out [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java). Try to implement a trivial extension on top of it, or help us port an existing plugin. You could also pick up one of the issues labeled "[good first issue](https://github.com/opensearch-project/opensearch-sdk-java/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)" in that project or start with one of the ideas we mentioned above. As always, please let us know how we can help by opening new issues or posting to [the forums](https://forum.opensearch.org/).
```
